File: macOS/MediaFlowMonitor/Sources/MediaFlowMonitor/Licensing/LicenseManager.swift
import Foundation

/// Stare de licențiere: probă gratuită 15 zile (Regula 3), apoi activare
/// prin serial (LicenseCore/Ed25519) generat manual din Furnizor.
/// Susținerea aplicației e o DONAȚIE, niciodată "preț"/"cumpără"/"vânzare".
enum LicenseState: Equatable {
    case trial(daysLeft: Int)
    case trialExpired
    case licensed(expiresAt: Int64) // 0 = lifetime
}

final class LicenseManager: ObservableObject {
    static let productID = "media-flow-monitor"
    private static let trialDays = 15

    @Published private(set) var state: LicenseState = .trial(daysLeft: LicenseManager.trialDays)

    private let defaults = UserDefaults.standard
    private let firstLaunchKey = "MediaFlowMonitor.firstLaunchDate"
    private let serialKey = "MediaFlowMonitor.serial"

    init() {
        if defaults.object(forKey: firstLaunchKey) == nil {
            defaults.set(Date(), forKey: firstLaunchKey)
        }
        refresh()
        Task {
            await RevocationCheck.shared.refresh(productIDs: [Self.productID])
            refresh() // reia decizia locală dacă revocarea tocmai a sosit online
        }
    }

    func refresh() {
        if let serial = defaults.string(forKey: serialKey),
           case .success(let payload) = LicenseCore.validate(serial: serial, expectedProductID: Self.productID),
           !RevocationCheck.shared.isRevoked(Self.productID) {
            state = .licensed(expiresAt: payload.expiresAt)
            return
        }

        let firstLaunch = defaults.object(forKey: firstLaunchKey) as? Date ?? Date()
        let daysUsed = Calendar.current.dateComponents([.day], from: firstLaunch, to: Date()).day ?? 0
        let daysLeft = Self.trialDays - daysUsed
        state = daysLeft > 0 ? .trial(daysLeft: daysLeft) : .trialExpired
    }

    /// Activează cu un serial primit de la furnizor (după donație, prin WhatsApp).
    func activate(serial: String) -> Result<Void, LicenseCore.ValidationError> {
        switch LicenseCore.validate(serial: serial, expectedProductID: Self.productID) {
        case .success:
            defaults.set(serial, forKey: serialKey)
            refresh()
            return .success(())
        case .failure(let error):
            return .failure(error)
        }
    }

    var machineIDDisplay: String { MachineID.display }

    var whatsAppActivationURL: URL {
        let text = "Salut! Vreau Lifetime Access pentru MediaFlow Monitor (donatie 7e). Machine ID: \(machineIDDisplay)"
        let encoded = text.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? text
        return URL(string: "https://wa.me/34643109970?text=\(encoded)")!
    }
}

File: Windows/MediaFlowMonitor/Licensing/LicenseManager.cs
using System;
using System.ComponentModel;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;
using Microsoft.Win32;

namespace MediaFlowMonitor.Licensing;

public enum LicenseKind { Trial, TrialExpired, Licensed }

public readonly record struct LicenseState(LicenseKind Kind, int DaysLeft = 0, long ExpiresAt = 0);

/// Port al LicenseManager.swift — probă gratuită 15 zile (Regula 3), apoi
/// activare prin serial (LicenseCore/Ed25519) generat manual din Furnizor.
/// Persistat în Registry (HKCU), la fel ca ThemeManager/CacheFolderLocator.
public sealed class LicenseManager : INotifyPropertyChanged
{
    public const string ProductID = "media-flow-monitor";
    private const int TrialDays = 15;
    private const string RegistryKey = @"Software\GDC\MediaFlowMonitor";

    public event PropertyChangedEventHandler? PropertyChanged;

    public LicenseState State { get => _state; private set { _state = value; PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(State))); } }
    private LicenseState _state;

    public LicenseManager()
    {
        using (var key = Registry.CurrentUser.CreateSubKey(RegistryKey))
        {
            if (key.GetValue("FirstLaunchDate") == null)
                key.SetValue("FirstLaunchDate", DateTime.UtcNow.ToString("O"));
        }
        Refresh();
        _ = Task.Run(async () =>
        {
            await RevocationCheck.Shared.RefreshAsync(new[] { ProductID });
            Refresh(); // reia decizia locală dacă revocarea tocmai a sosit online
        });
    }

    public void Refresh()
    {
        using var key = Registry.CurrentUser.OpenSubKey(RegistryKey);
        var serial = key?.GetValue("Serial") as string;

        if (!string.IsNullOrEmpty(serial))
        {
            var (payload, error) = LicenseCore.Validate(serial, ProductID);
            if (error == null && payload is { } p && !RevocationCheck.Shared.IsRevoked(ProductID))
            {
                State = new LicenseState(LicenseKind.Licensed, ExpiresAt: p.ExpiresAt);
                return;
            }
        }

        var firstLaunchStr = key?.GetValue("FirstLaunchDate") as string;
        var firstLaunch = DateTime.TryParse(firstLaunchStr, null,
            System.Globalization.DateTimeStyles.RoundtripKind, out var parsed) ? parsed : DateTime.UtcNow;
        int daysUsed = (int)(DateTime.UtcNow - firstLaunch).TotalDays;
        int daysLeft = TrialDays - daysUsed;
        State = daysLeft > 0
            ? new LicenseState(LicenseKind.Trial, DaysLeft: daysLeft)
            : new LicenseState(LicenseKind.TrialExpired);
    }

    /// Activează cu un serial primit de la furnizor (după donație, prin WhatsApp).
    public LicenseCore.ValidationError? Activate(string serial)
    {
        var (_, error) = LicenseCore.Validate(serial, ProductID);
        if (error != null) return error;

        using var key = Registry.CurrentUser.CreateSubKey(RegistryKey);
        key.SetValue("Serial", serial);
        Refresh();
        return null;
    }

    public string MachineIDDisplay => MachineID.Display;

    // Preț dinamic (Regula 27) - vezi PricingChecker. Fail-open pe 7 €
    // (valoarea hardcodata anterior) daca pricing.json nu e accesibil.
    public string WhatsAppActivationUrl
    {
        get
        {
            var priceText = PricingChecker.Shared.DisplayText;
            var text = $"Salut! Vreau Lifetime Access pentru MediaFlow Monitor (donatie {priceText}). Machine ID: {MachineIDDisplay}";
            return $"https://wa.me/34643109970?text={Uri.EscapeDataString(text)}";
        }
    }

    public string StatusText => State.Kind switch
    {
        LicenseKind.Trial => $"Probă gratuită: {State.DaysLeft} zile rămase",
        LicenseKind.TrialExpired => "Probă expirată — activare necesară",
        LicenseKind.Licensed => State.ExpiresAt == 0 ? "Licențiat (Lifetime)" : "Licențiat",
        _ => "",
    };
}
